engine/src/main/java/org/terasology/network/BroadCastServer.java

```diff
@@ -0,0 +1,135 @@
+    }
+
+    public void setTurnOnBroadcast(boolean turnOnBroadcast) {
+        BroadCastServer.turnOnBroadcast = turnOnBroadcast;
+    }
+
+    public void startBroadcast() {
+        try {
+            sendBroadCast();
+            service = Executors.newSingleThreadScheduledExecutor();
+            Runnable runnable = new Runnable() {
+                public void run() {
+                    logger.info("Listening to broadcast");
+                    try {
+                        listenToBroadCast();
+                    } catch (Exception e) {
+                        logger.error("Broadcasting has been interrupted" + e.getMessage());
+                    }
+                }
+            };
+            service.scheduleAtFixedRate(runnable, 0, 300, TimeUnit.SECONDS);
+        } catch (Exception e) {
+        logger.info("Broadcasting has been interrupted " + e.getMessage());
+        e.printStackTrace();
+        }
+    }
+
+    public void stopBroadCast() {
+        logger.info("Shutting down BroadCasting");
+        service.shutdown();
+    }
+
+    private void listenToBroadCast() throws Exception {
+        byte[] buffer = discoveryRequest.getBytes(StandardCharsets.UTF_8);
+        DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
+        getReceiveSocket().receive(packet);
+        logger.info("Discovery package received! -> " + packet.getAddress() + ":" + packet.getPort());
+
+        //Validate the data sent.
+        String data = new String(packet.getData()).trim();
+        if (data.equals(discoveryRequest)) { // validate command
+            // Send response
+            byte[] response = new byte[discoveryResponse.length()];
+            DatagramPacket responsePacket = new DatagramPacket(response, response.length, packet.getAddress(), packet.getPort());
+            getReceiveSocket().send(responsePacket);
+            logger.info("Response sent to: " + packet.getAddress() + ":" + packet.getPort());
+        } else {
+            logger.info("Error, not valid package!" + packet.getAddress() + ":" + packet.getPort());
+        }
+    }
+
+    private DatagramSocket getReceiveSocket()  {
+        try {
+            if (receiveSocket == null) {
+                receiveSocket = new DatagramSocket(8002,
+                    InetAddress.getByName("0.0.0.0")); // 0.0.0.0 for listen to all ips
+                receiveSocket.setBroadcast(true);
+            }
+        } catch (Exception e) {
+            logger.error("Broadcast Exception Encountered" + e.getMessage());
+        }
+        return receiveSocket;
+    }
+
+    private DatagramSocket getSendSocket() {
+        try {
+            if (sendSocket == null) {
+                sendSocket = new DatagramSocket(8001, InetAddress.getLocalHost());
+                sendSocket.setBroadcast(true);
+            }
+        } catch (Exception e) {
+            logger.error("Broadcast Exception Encountered" + e.getMessage());
+        }
+        return sendSocket;
+    }
```

Comment:
We won't be needing a new socket right? Because we would be using the same one to send the request. Do you think there is a better way to do this?

Reply:
I'm would do separate socket intialization and getters, because it can throw exception.
If it will throw exception, then broadcast feature cannot will work properly at least.
If you will use one class instance - you use one socket.
Additional, You can split client and server functionality

Reply:
In my opinion this code should catch broadcast errors if the sockets are null or the ports are unavailable so that the broadcast can be terminated immediately. I do not think we need more modularity  in terms of client and server functionality because this offers us the best error handling.